Substantial Refactor

makinde opened the pull request:
This started as a much smaller effort to fix the authorization check for the `save` hook, since I'm pretty sure that wasn't working. Then I discovered that the `vm` variable name meant at least 3 different things, so I renamed that in a bunch of places. Then I just went for it.

Here are the major changes:
- Fixing the `save` hook so it should work now
- Adding the ability to specify that a type (or all) authorization is mandatory (and causing an error if an authLevel can't be determined).
- Adding a new way to specify/determine the authLevel. This involves using a method on the schema that can take in a specific document and a payload from the query to determine a specific authLevel. The main use case is when the authLevel depends on the document (if you want to check if a specific user owns a specific document, in which case you'd return 'OWNER' as the authLevel, otherwise 'VIEWER'). Since this requires the document, it is done as a post-save hook when doing a find query and it doesn't work for updates/deletes.

I then did a couple passes and simplifying the logic.

Lots of testing will be needed here, but I wanted to share this sooner rather than later to see if this had a shot at being upstreamed and to get some feedback.

makinde committed:
Substantial Refactor

This started as a much smaller effort to fix the authorization check for the `save` hook, since I'm pretty sure that wasn't working. Then I discovered that the `vm` variable name meant at least 3 different things, so I renamed that in a bunch of places. Then I just went for it.

Here are the major changes:
- Fixing the `save` hook so it should work now
- Adding the ability to specify that a type (or all) authorization is mandatory (and causing an error if an authLevel can't be determined).
- Adding a new way to specific/determine the authLevel. This involves using a method on the schema that can take in a specific document and a payload from the query to determine a specific authLevel. The main use case is when the authLevel depends on the document (if you want to check if a specific user owns a specific document, in which case you'd return 'OWNER' as the authLevel, otherwise 'VIEWER'). Since this requires the document, it is done as a post-save hook when doing a find query and it doesn't work for updates/deletes.

I then did a couple passes and simplifying the logic.

Lots of testing will be needed here, but I wanted to share this sooner rather than later to see if this had a shot at being upstreamed and to get some feedback.

chrismatix commented on index.js:
```diff
@@ -27,58 +33,72 @@ module.exports = function (schema) {
+            .map(function(level) {
+                return schema.permissions[level] && schema.permissions[level][action];
+            })
+            .faltten()
```
typo, should be `flatten`

chrismatix commented:
Hey, thanks a lot for taking the time to do this. There is a lot of stuff that needs to be refactored here and you did an amazing job already.

I also really like the conditional auth level that you introduced. From where I am standing this is perfectly fine. Spotted one minor typo and of course for merge I would like to see the new cases covered by tests.

Once you are done, could you also increase the version by a minor and maybe create a changelog?

Again, thanks a lot!

makinde committed:
Fixing typo

makinde commented:
@xChrisPx Thanks for taking a look! Cool, I'll keep moving forward in this direction.

- **Testing**: Any guides/pointers on running the tests locally? The travis build is failing, but there's not much in the way of helpful error messages.
- **Changelog**: Yup, can look into this, and the minor version bump.

chrismatix commented:
@makinde yes, I added an `npm test` script in the last release. You only need a local mongod running on the standard port. Existing tests are in the /test folder. For the changelog what you have written above should be fine.

makinde committed:
Getting the current tests to pass

These aren't exhaustive, but they are a nice baseline.

makinde commented:
Ah man, been hacking away at this for the last couple days. For context, I'd like to use this in a production app fairly soon, so I want to make sure it's good to go.

Over the last couple days, I've chipped away at a lot of issues I've uncovered.
- The module doesn't cover all the possible hooks people might use to edit docs (e.g. there's no hook for `document.remove()`.
- The fields in the permissions object seem strange to me. `write`, `save`, and `update` are pretty confusing. I'm changing them to `read` & `write` (both are lists of fields) and then `create` and `remove` (booleans).
- a bunch more small tweaks.

This is basically writing a new library, and I don't think it'll be super backwards compatible. I'm new to OSS, and not super sure what the right approach is here. I could keep pushing forward here and submit a PR, but that feels odd considering every line is gonna change. I could also hard fork the project and create a new name on npm for the package. Is that an okay thing to do?

@xChrisPx , if you think that latter is okay, I'd be happy to keep working on this, get the code reviewed here, and then keep working with you on that new project. Would love to have you as a co-maintainer there. Also, any ideas on names?

makinde committed:
Phew, more overhaul and changes

- Global enable/disable. The combination of the defaults list and being off by default was hard to reason about
- Authorization is now mandatory by default
- Change permissions settings to read, write, create, and remove.
- Added Linting
- Much more edge case handling for find and updates

makinde commented:
@xChrisPx Just took another pass. There are tons of edge cases to throughly test here, and I'm not sure if I'm headed down the right path. Any chance you have the bandwidth to help flesh out more tests?

I'm pretty happy with the structure of the code at this point. It seems like `read`, `write`, `create`, `remove` are the right permissions to go with. Any thoughts on this breaking change?

makinde committed:
Higher node version to make things muuuuch easier

makinde committed:
Cleaning up todo's

chrismatix commented:
Hey @makinde,

this is also what brought me to the project. Also my first real oss experience. I think the proper way, in order to keep this easily accessible would be to go *2.0.0* (compatability breaking changes) and add a small transition guide. But if you want to start a new one, I am game. There is a lot of nice stuff that I want to add like Typescript. 

I will have time to give it a thorough review by tomorrow, if that is ok? However if you need this urgently in production I can quickly push this as 2.0.0 and we can deploy it as an alpha so that nobody is impacted by this change yet.

chrismatix commented:
However on my first glance: Amazing work!

makinde committed:
Making the helpers work with virtuals

makinde committed:
Updating tests to use async/await and also making `find` hooks work!

makinde committed:
Revert version change

Doing this so we can use the np module to do a beta release

brysgo commented on lib/helpers.js:
```diff
@@ -0,0 +1,49 @@
+  return _.chain(authLevels)
+    .map(level => schema.permissions[level][action])
+    .flatten()
+    .filter(path => schema.path(path) || schema.virtualpath(path)) // make sure the fields are in the schema
```
Is this line really needed? I think it is removing nested objects from the list even though they are checked (breaks writing nested objects)

makinde replied:
Ah, that's a good case to consider. To be clear, you're talking about a schema like this?

```
const schema = Schema({
  foo: {
    bar: String
  }
});
```

And you're saying that you won't be able to set permissions like this?
```
schema.permissions = {
  'foo.bar': { read, write, remove }
};
```

I was torn on adding that extra filter. by having it, the rest of the code was able to not have to worry about bogus paths coming back. Is there a way to validate that paths you are mentioning based on the schema?

makinde committed:
Updating readme for new library logic

makinde committed:
Lots of improvements...

- Adding a bunch more tests. I feel like the coverage is there and this probably mostly works now
- Adding a `permissions` option to find queries. If that's set to try, permissions will be injected into the returned object.
- Readme is sorta decent, but not really.

makinde committed:
removing local db start/stop code

makinde committed:
fixing package.json

makinde commented:
@xChrisPx I finally had some time to circle back to this and pushed it forward. I've added test so I'm pretty sure this works in the typical cases. I've left a bunch of TODO's in there to cover more corner cases of mongoose, but I think this is ready for a release candidate.

Do you know how to set that up and do you have the permissions? If not, I can start to tinker with this on my app by loading it directly from my fork, but it would be good to get it flowing through NPM soon.

makinde committed:
shuffling around some TODO's

chrismatix commented:
@makinde great. Then as I said in my previous comment, I would publish this as **2.0.0**, because as you pointed out earlier, this is almost a completely new library. Can you update the package.json?

makinde committed:
bumping version

makinde commented:
Sweet, just bumped it to `2.0.0-alpha.1`. Thanks for the help. Once we get this up, I can create a bunch of issues for what needs to be done get this to a full `2.0.0`. Basically more tests and more mongoose edge case handling.

brysgo commented:
Awesome! (we should just throw an error if you try and update without an authLevel re: edge cases)

makinde commented:
@brysgo Yup. You get a `PermissionDeniedError` if you don't pass an `authLevel`. The edge cases are things like the `overwrite` option in `Model.update()` that can let you overwrite documents with and arbitrary object (e.g. `{}`) from an update query. That command would end up removing a bunch of keys, and we'd need to add logic to `mongoose-authorization` to figure out which keys are going to be updated. I left a TODO in the codebase for that.

FWIW, My use case only involves using this library with a `getAuthLevel` method in the schema. I actually never pass in in an authLevel directly. I left that in there mainly for backward compatibility and because it seems like it actually might be more common for people. But my authLevels are always dependent on the logged-in user and the given document, so I push the logic into `getAuthLevel`. So you'll see that my focus is on paths that hit that flow, and basically the entire `update` flow doesn't have access to the document, so my code never uses that command, and then it's the least built out in here :P

makinde commented:
Also, thanks for all the help in getting this out. This has been a pretty cool, if not painful at times, project. 🤝 I never thought I'd be this far down the rabbit hole of mongoose internals.

brysgo commented:
Haha, so awesome, thanks @makinde -- I was also trying to use `getAuthLevel` exclusively but updates present a challenge. Unless you fixed it?

brysgo commented on lib/helpers.js:
```diff
@@ -0,0 +1,144 @@
+
+  return _.chain(authLevels)
+    .flatMap(level => schema.permissions[level][action])
+    .filter(path => schema.path(path) || schema.virtualpath(path)) // ensure fields are in schema
```
This line breaks access to subdocuments for me (posted a failing test on the forked repo), is there a security reason that we can't access things that aren't in the schema?

makinde replied:
Let's chat about this in a new issue. I just created #8 for us to discuss.

makinde commented:
Yeah, @brysgo, I decided in my codebase not to use updates at all. We always load the object, modify, and then save. We already worked that way to make sure validation works (some of our validators check that a group of fields are okay together). Also, most of our permissions need a copy of the object (is this person the owner? Is this person attending the event?), so that's another reason we couldn't use `Model.update()`.

brysgo commented:
@makinde - I'm cool with that, as long as there is an error when someone tries to update. I don't even want to get into my `getAuthLevel` if it is an update because it is confusing and potentially insecure.

makinde commented:
Moved the discussion to #9 @brysgo

makinde committed:
Removing extra code and adding a test

makinde committed:
Adding test for disabling authorization on lower levels of population

makinde committed:
Fixing #8 - Support nested paths

https://github.com/352Media/mongoose-authorization/issues/8

Phew, now we can support nested paths, at least on find queries. See tests for more detail.

makinde commented:
Hey, can we merge this? It'd be good to start doing real PR's and not merging directly into this branch all the time. Happy for that step even if we don't publish to npm right away. @xChrisPx @dvideby0
